### ISSUE TITLE
Support API Key Auth

### ISSUE THREAD

**Issue by arkodg** [envoyproxy/gateway#2630]
> @zhaohuabing can we reuse the same Envoy Filter for API key auth ? 

The problem is, unlike HTTP Basic Auth, API key auth is much more flexible, the key can be in the header, path, or even body. I think we will need a dedicated envoy filter in Envoy if we want to support API key auth, we can follow the open API specs: https://swagger.io/docs/specification/authentication/api-keys/#:~:text=API%20keys%20are%20supposed%20to,mechanisms%20such%20as%20HTTPS%2FSSL.

_Originally posted by @zhaohuabing in https://github.com/envoyproxy/gateway/issues/1660#issuecomment-1814974031_

**Comment by arkodg** [comment]
blocked on upstream implementation in envoy

**Comment by zufardhiyaulhaq** [comment]
@arkodg do you know the upstream implementation envoy tracking?

**Comment by zufardhiyaulhaq** [comment]
I think this can be easily implemented by header or query matching right?

**Comment by arkodg** [comment]
cc @zhaohuabing

**Comment by zhaohuabing** [comment]
@arkodg  @zufardhiyaulhaq It should be a simple Envoy Filter. I can pick it up when I get some bandwidth.

**Comment by missBerg** [comment]
API Key Auth is also often related to a set of APIs that is accessible with that Key. Often stored elsewhere to determine if the resource (API) being accessed is allowed to be accessed with this Key.

What is the main purpose of this feature? To capture API key values for rate limiting only? Or to do AuthZ enforcement?

And how would we differentiate dynamically if an API key is subscribed to a higher rate limit than another API key?

**Comment by gecube** [comment]
@missBerg it is very essential feature for us stopping the adoption of envoy gateway. The primary idea is to use it as AuthZ. Regarding rate limits - we can support different scenarios. Probably starting from separate rate limit for each API key. 

> And how would we differentiate dynamically if an API key is subscribed to a higher rate limit than another API key?

It should not be an issue.

When the basic rate limiting would be implemented, complex scenarios could be considered (like common rate limit for API keys from one user).

**Comment by arkodg** [comment]
looks like the upstream feature is in, is this something we can get in for v1.3 @sanposhiho ?

**Comment by sanposhiho** [comment]
Yeah, how do we deal with the compatibility between gateway and envoy? Can we just introduce any latest features immediately or do we have to wait for envoy to cut a new release?

**Comment by sanposhiho** [comment]
Okay, I found:
https://gateway.envoyproxy.io/news/releases/matrix/

So, we can just start this implementation now, do I understand right?

**Comment by arkodg** [comment]
yeah @sanposhiho, gateway `main` is based off proxy `main`

**Comment by arkodg** [comment]
you'll also need to `go getgithub.com/envoyproxy/go-control-plane@<latest>`

**Comment by BryanFauble** [comment]
+1 to this feature/issue. This is an important feature for me. My use case:

I have OpenTelemetry data that I am ingesting from several sources (https://github.com/Sage-Bionetworks-Workflows/eks-stack/blob/ca523c66abc8d9b91e796c838f5d3c24784048cc/modules/signoz/main.tf#L252-L278 & https://github.com/Sage-Bionetworks-Workflows/eks-stack/blob/ca523c66abc8d9b91e796c838f5d3c24784048cc/modules/signoz/resources-otel-ingress/security-policy.yaml#L1-L13). I am currently enforcing that all incoming requests provide a JWT bearer token, however, I am limited with a monthly quota for the number of tokens that can be created per month via Auth0.

If I can swap to using an API key instead then I will be able to on-board more projects without the concern of filling up my monthly quota. This is also combined by the fact that the AWS supported OTEL collectors (For ECS and Lambda functions) have not implemented https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/extension/oauth2clientauthextension
